policy: T5069: large-community-list regex validator disallows whitespace

talmakion opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
NOTE: Potentially hijacking an old ticket with a similar problem. 

T5816 improved 1.4/1.5's strict validation to allow full regex on a single large-community tuple. However, it is common to match multiple components of a large-community string via regex in earlier VyOS versions (not to mention Cisco IOS). T5816 is too strict about the form of the regex pattern, does not allow the common-separator wildcard '_' and whitespace, breaking old 1.3 configs and preventing simple pattern logic being used to match multiple communities. 

To resolve T5069's concerns and my own, this patch attempts to:
 * Perform basic, easy sanity checks that should always pass:
   * Contain at least one well-formed tuple
   * Permit only specific characters valid in community strings and POSIX regex
   * Rejecting obviously incorrect regular expressions, treating Python regex as a superset of POSIX
 * Run a final "sanity check" regex across the top, issue a warning if the overall pattern is not considered well formed

"Well formed" in this case would be one or more community segments roughly fitting the original check, separated by proper whitespace. "1:1:12:2:2" might be a mistake for "1:1:1 2:2:2", but there are still instances of useful patterns not following this form. 

Since there's no mechanism for providing warning feedback from a validator and validator output is thrown away unless an error, I've wrapped this in the console_warning() function and it just prints directly to the TTY. If this seems like overkill, if it could be more generally useful, or if there's a suitable mechanism I missed - I imagine validation success/warning/failure feedback could be useful at API level, not just CLI - let me know and I'll update it to match. 

If we retain the warning, I'll likely issue a followup patch for the documentation explaining the specific validation logic and what to check when the warning pops up. 

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
* https://vyos.dev/T5069
* https://vyos.dev/T5816 - only for reference

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
# Example tests via CLI:
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1"

  
  
  
  Malformed large-community-list
  Value validation failed
  Set failed

[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1"
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:12:2:2"
WARNING: Regex does not follow expected form and may not match as expected.
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1 2:2:2"
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1_.*_2:2:2"
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1_.*_2:2:2:"
WARNING: Regex does not follow expected form and may not match as expected.
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1_.*(_2:2:2)"
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1_.*(_2:2:2))"

  
  
  
  Malformed large-community-list
  Value validation failed
  Set failed

[edit]
```

I've checked that these compound matches are accepted by FRR and work in a production BGP environment as the older versions of FRR in 1.3 did. 

Smoketests, test_policy.py:
```
Ran 28 tests in 375.194s

OK
```

test_protocols_bgp.py:
```
Ran 30 tests in 438.052s

OK
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

dmbaturin commented on src/validators/bgp-large-community-list:
```diff
@@ -17,18 +17,47 @@
-pattern = '(.*):(.*):(.*)'
-allowedChars = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '+', '*', '?', '^', '$', '(', ')', '[', ']', '{', '}', '|', '\\', ':', '-' }
+def console_warning(txt: str):
+    with open('/dev/tty', 'w') as tty:
```
We already have a thing specially for this purpose: [`vyos.base.Warning`](https://github.com/vyos/vyos-1x/blob/current/python/vyos/base.py#L47). Warnings also should be printed in non-interactive script calls as well, for testing, for example.

talmakion replied:
Cheers for pointing me in that direction - I wasn't sure if there was already a generic mechanism for warnings. 

The problem is that Warning() output is still eaten by the validation caller, same as any normal output on stdout/stderr, unless we exit non-zero. 

Change console_warning to a Warning instance, no more warning output:
```
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:12:2:2"
[edit]
```

Running it raw:
```
vyos@TEST-VYOS-LEFT# python3 /usr/libexec/vyos/validators/bgp-large-community-list "1:1:12:2:2"

WARNING: Regex does not follow expected form and may not match as
expected.
```

Change the sys.exit(0) to sys.exit(1) (it's definitely hitting it):
```
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:12:2:2"

  
  WARNING: Regex does not follow expected form and may not match as
  expected.
  
  
  Malformed large-community-list
  Value validation failed
  Set failed

[edit]
```

Not sure what the neatest way forward would be: 
* I can customise the Warning.print() handling for this specific use, also output to /dev/tty
* If it's useful elsewhere, add a flag to enable output to /dev/tty in the base implementation
* Get vyos1x-config to still display output without failing when flagged to do so, maybe a different exit code? Does anything else interact with the validators directly?

I don't particularly like the forced direct-to-TTY but it was the only way I could find to actually produce output for the user to see.

talmakion replied:
I've gone with using exit codes from the validator to signal there's non-failing output and changed the warning to use the normal wrapper.

I've raised a [PR](https://github.com/vyos/vyos-utils/pull/32) to have validate_value handle this correctly. 

Combined, this works pretty close to what I was after:
```
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:1 2:2:2"
[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1:12:2:2"
  
  WARNING: Regex does not follow expected form and may not match as
  expected.

[edit]
vyos@TEST-VYOS-LEFT# set policy large-community-list TEST rule 20 regex "1:1"
  
  
  Malformed large-community-list
  Value validation failed
  Set failed

[edit]
```

talmakion replied:
Following on the comments on the [other PR](https://github.com/vyos/vyos-utils/pull/32), I've updated this to include a UserMessage with redirection capability and a validator-script specific warning wrapper to send everything to stderr.

I've made the assumption you probably don't want specific warning wrappers in vyos.base, and that a specific wrapper is better than a generic WarningStderr() or similar. 

I haven't had a chance yet to exercise other validators to find any that were chatty on stderr - I vaguely recall some were in older VyOS releases - but if this basic form suits, I can give them a look over.

dmbaturin reviewed (changes requested):
The fix looks sensible but please use the standard warning mechanism instead of a custom alternative.

talmakion commented on src/validators/bgp-large-community-list:
```diff
@@ -16,19 +16,45 @@
+        sys.exit(1)
+
+    # Simple guard against invalid community & 1003.2 pattern chars
+    if not set(value).issubset(allowedChars):
```
This is already present.

talmakion commented on src/validators/bgp-large-community-list:
```diff
@@ -16,19 +16,45 @@
+        # If you know what you're doing, you should be able to do it.
+        # It is, however, very easy to over- or under-restrict compound regex 
+        # logic against communities and ASNs, especially if they change. 
+        # Best practice: stick with basic patterns, mind your wildcards and whitespace.
```
This is no longer part of the change.

talmakion commented:
Unsure why the smoketest failed - it is UserMessage/Warning-related, but I can't get it to fail in local. Comparing a vanilla VyOS with an updated one, manual smoketest runs and direct config sets work as expected. 

I did muff the previous force-push so I've just tweaked minor Ruff lint fails and some naming that didn't track changes in vyos.base to trigger another smoketest run. 

Edit: It was a vyos-configd interaction, need to use runtime values of sys.stdout.

Commit:
policy: T5069: large-community-list regex validator should allow whitespace
* Re-introduce the whitespace/pattern matches ' ' and '_' as allowed
* Perform a general Python regex validity check (not 100% 1003.2, but in combination
  with allowedChars, pretty close)
* Introduce a warning against potentially malformed or over-complex patterns,
  but leave it up to the user to resolve - there are plenty of useful
  expressions we cannot validate easily

talmakion commented:
Updated version: split off the warning-logic into conf verify and dropped output changes for validators. 

Re-ran smoketests for policy OK.

sever-sever reviewed (approved):
Fixes BGP large-community regex
Smoketest passed. Approved.

dmbaturin reviewed (approved):
The warning is now in the config scripts. No objections against the logic.
